src/control/destination/background/destination_set.rs

```diff
@@ -0,0 +1,352 @@
+use std::{
+    collections::HashMap,
+    fmt,
+    iter::IntoIterator,
+    net::SocketAddr,
+    time::{Instant, Duration},
+};
+
+use futures::{Async, Future, Stream,};
+use tower_h2::{BoxBody, HttpService, RecvBody};
+
+use linkerd2_proxy_api::{
+    destination::{
+        update::Update as PbUpdate2,
+        WeightedAddr,
+    },
+    net::TcpAddress,
+};
+
+use super::super::{Metadata, Responder, Update};
+use control::{
+    cache::{Cache, CacheChange, Exists},
+    remote_stream::Remote,
+};
+use dns::{self, IpAddrListFuture};
+use telemetry::metrics::DstLabels;
+use transport::{tls, DnsNameAndPort};
+use conditional::Conditional;
+
+use super::{DestinationServiceQuery, UpdateRx};
+
+/// Holds the state of a single resolution.
+pub(super) struct DestinationSet<T: HttpService<ResponseBody = RecvBody>> {
+    pub addrs: Exists<Cache<SocketAddr, Metadata>>,
```

Comment:
I'd like to make some of the fields on `DestinationSet` not be `pub`, and have `Background` interact with them using methods with more descriptive names. Planning on making that change in a follow-up branch.